[TypeChecker] Implement distributed computed properties

xedin opened the pull request:
Add support for parsing `distributed var <name> : <type> { ... }` declarations and
type-checking them by injecting distributed thunks during Sema.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Resolves rdar://91283164

ktoso committed:
wip

ktoso committed:
towards distributed getters

xedin committed:
[Distributed] Synthesize thunks for distributed computed properties

xedin committed:
[Distributed] Remove commented out code and print statements

xedin committed:
[Distributed] SILGen: Remove `isDistributed` flags from accessor code

xedin committed:
[TypeChecker] Distributed: Verify properties before attempting to synthesize a thunk

xedin committed:
[CSApply] Distributed: Inject distributed thunk when necessary

Replace distributed member references with distributed thunks
when access happens outside of distributed actor context. This
significantly simplifies distributed compute properties implementation.

ktoso committed:
fix checkDistributedTargetResultType to also return when not diagnosing

ktoso commented on lib/AST/DistributedDecl.cpp:
```diff
@@ -1333,6 +1333,20 @@ AbstractFunctionDecl *ASTContext::getRemoteCallOnDistributedActorSystem(
+  // Only get-only 'distributed' computed properties are considered valid.
+  if (isStatic() || isLet() || hasStorageOrWrapsStorage() ||
+      getWriteImpl() != swift::WriteImplKind::Immutable)
+    return nullptr;
```
We can remove this since https://github.com/apple/swift/pull/59481/commits/4cf402bd031b8b82cc1bd6f61aae8dd97cf06b88 fixes the checking to bail out when the prop is invalid, so we don't need to replicate the logic 👍

xedin replied:
Sounds good!

ktoso commented on lib/SILGen/SILGenDistributed.cpp:
```diff
@@ -212,6 +212,7 @@ void SILGenFunction::emitDistActorIdentityInit(ConstructorDecl *ctor,
   initializeProperty(*this, loc, borrowedSelfArg, var, temp);
 }
 
+// TODO(distributed): rename to DistributedActorID
```
Yes, thanks!

ktoso committed:
[Distributed] adjust test to pass with less accurate diagnostics; while we work on getting them back

ktoso commented:
Okey so I made the tests pass but... We're missing the info that "this was a distributed thunk" and I'm sorry but I forgot what we agreed upon -- are we doing the new attribute to mark those? It would honestly be the simplest solution and also useful to inspect those ASTs then, and we might need it for the avoiding double-synthesis anyway, right @xedin?

In any case this will now pass, and let's follow up fixing the FIXME's in https://github.com/apple/swift/pull/59481/commits/1178f624fb63858cf5028a656b8c0b20f44d2530 by introducing an attr or what other idea you had there :+1:

ktoso commented:
@swift-ci please test

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -520,6 +520,11 @@ namespace {
+    /// Determine whether the given reference on the given
+    /// base has to be replaced with a distributed thunk instead.
+    bool requiresDistributedThunk(Expr *base, SourceLoc memberLoc,
+                                  ConcreteDeclRef memberRef);
```
Cool 👍 I think this will make implementing the known to be local stuff very simple... since we'll add the logic in here to say "well, actually no, we know it is local -> no thunk" ❤️

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -1615,6 +1620,66 @@ namespace {
+        if (auto *FD = dyn_cast<FuncDecl>(decl)) {
+          thunkDecl = FD->getDistributedThunk();
+        } else {
+          thunkDecl = cast<VarDecl>(decl)->getDistributedThunk();
```
maybe 
```suggestion
          } else if (let var cast<VarDecl>(decl)) { 
            thunkDecl = var->getDistributedThunk();
          } else llvm_unreachable...
```

? Just in case I guess...

xedin replied:
The validity is checked by `requiresDistributedThunk` so I don't really see a reason to double-check here, if anything cast is going to fail.

ktoso replied:
Gotcha, thanks!

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -1615,6 +1620,66 @@ namespace {
+        declRefExpr->setFunctionRefKind(choice.getFunctionRefKind());
+        declRefExpr->setType(thunkType);
+
+        cs.cacheType(declRefExpr);
```
oh that's new to me, why do we do this?

xedin replied:
Are you talking about `cacheType` specifically? That's used to let `finishApply` know what the type was for the newly created node, since all the expressions involved in solving should have their types cached.

ktoso replied:
Yeah, that -- I see, thanks :)

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -1615,6 +1620,66 @@ namespace {
+        }
+
+        return thunkApply;
+      }
```
Sweet, this is very cool that we're able to just handle it all in this place.

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -202,23 +201,50 @@ deriveBodyDistributed_thunk(AbstractFunctionDecl *thunk, void *context) {
+    auto returnLocalPropertyAccess = new (C) ReturnStmt(sloc, localPropertyAccess, implicit);
+    localBranchStmt =
+        BraceStmt::create(C, sloc, {returnLocalPropertyAccess}, sloc, implicit);
+  } else {
```
<3

ktoso commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -431,10 +431,19 @@ static bool varIsSafeAcrossActors(const ModuleDecl *fromModule,
+        // which are allowed, but subject to the same implicit throws/async
+        // as all other distributed function declarations.
+        if (var->isDistributed()) {
+          return true;
```
right 👍

ktoso reviewed (approved):
Thank you, looking awesome.

The only follow up is that I forgot what we agreed to do about the thunks (too little sleep -_-) -- I'd be ok with the attribute to detect them, but do we want to do something else?

xedin committed:
[AST] Remove `shouldApplyLookDistributedThunk` flag from `LookupExpr`

The underlying mechanism has changed so the flag is no longer required.

xedin commented:
>  are we doing the new attribute to mark those? It would honestly be the simplest solution and also useful to inspect those ASTs then, and we might need it for the avoiding double-synthesis anyway, right @xedin?

Yes, that's is correct.

xedin committed:
[AST] ApplyExpr: Rename `shouldApplyDistributedThunk` to `usesDistributedThunk`

Since the thunk is applied to expression during solution application,
the use of this flag has shifted towards stating the fact.

xedin committed:
[CSApply] Fix distributed thunk requirement check to handle non-parameter/capture base

If distributed memeber is referenced on a base that represents
a result of some expression which is not a variable, parameter,
or a capture it should always be dispatched via a thunk.

xedin committed:
[TypeChecker] Concurrency: turn `checkDistributedAccess` into validity check

All the heavy lifting of checking whether reference throws and
requires dispatch through a thunk is done during type-checking,
 so all `checkDistributedAccess` needs to do is determine whether
access is distributed or not.

xedin committed:
[Distributed] NFC: remove a couple of unused variables from sythensis code

xedin committed:
[Distributed] Sema: Add a new distributed-thunk attribute

The attribute comes handy during solution application to
determine whether the call is using a distributed thunk.

xedin commented:
@ktoso Tests have been fixed and new attributed simplified checking some what as well, so the changes are complete now.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -1615,6 +1616,70 @@ namespace {
+        if (isa<VarDecl>(decl)) {
+          auto *thunkCall = CallExpr::createImplicit(
+              context, thunkApply,
+              ArgumentList::forImplicitUnlabeled(context, /*args=*/{}));
```
I forgot that there is `CallExpr::createImplicitEmpty`, going to change it now.

xedin commented:
@swift-ci please test

ktoso commented:
Awesomeness, thank you Pavel :-)

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -7790,14 +7790,14 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+    if (isa<SelfApplyExpr>(apply->getFn())) {
+      auto *FD = dyn_cast<FuncDecl>(callee.getDecl());
+      if (FD && FD->isDistributedThunk()) {
         apply->setImplicitlyThrows(true);
```
This is very nice and simple now, came togeher nicely :)

ktoso commented:
I'll pick it over to 5.7

DougGregor commented on lib/AST/Decl.cpp:
```diff
@@ -7781,6 +7781,10 @@ bool AbstractFunctionDecl::isSendable() const {
 }
 
+bool AbstractFunctionDecl::isNonisolated() const {
+  return getAttrs().hasAttribute<NonisolatedAttr>();
```
This isn't a semantic definition of "nonisolated" (one should check `getActorIsolation` instead), so I'd prefer that we not have this function. Fortunately, it's unused, so we should remove it.

ktoso replied:
Ah I see, cool will remove 👍

ktoso replied:
https://github.com/apple/swift/pull/59634
